src/hotspot/share/opto/predicates.cpp

```diff
@@ -147,3 +148,331 @@ Node* PredicateBlock::skip_regular_predicates(Node* regular_predicate_proj, Deop
+  //     [top, i]
+  // where
+  //     top->in(i) = old_parent.
+  // We rewire the top node to the new parent such that:
+  //     top->in(i) = new_parent
+  void rewire_node_on_top_to(Node* new_parent) {
+    Node* top = _stack.top();
+    assert(is_cloned_node(top) && !top->is_Opaque1(), "must be cloned node on chain to OpaqueLoop*Node (excluded)");
+    const uint index_to_old_parent = _stack.node_index_to_previously_visited_parent();
+    top->set_req(index_to_old_parent, new_parent);
+  }
+
+ public:
+  CloneTemplateAssertionPredicateBool(BoolNode* template_bool, Node* ctrl_for_clones, PhaseIdealLoop* phase)
+      : _stack(template_bool),
+        _phase(phase),
+        _index_before_cloning(phase->C->unique()),
+        _ctrl_for_clones(ctrl_for_clones)
+  DEBUG_ONLY(COMMA _found_init(false)) {}
+
+  // Look for the OpaqueLoop* nodes to transform them with the strategy defined with 'transform_opaque_loop_nodes'.
+  // Clone all nodes in between.
+  BoolNode* clone(TransformOpaqueLoopNodes* transform_opaque_loop_nodes) {
+    Node* current;
+    while (_stack.is_not_empty()) {
+      current = _stack.top();
+      if (current->is_Opaque1()) {
+        transform_opaque_loop_node(current, transform_opaque_loop_nodes);
+        pop_transformed_opaque_loop_node();
+      } else if (!_stack.push_next_unvisited_input()) {
+        pop_node();
+      }
+    }
+    assert(current->is_Bool() && current->_idx >= _index_before_cloning, "new BoolNode expected");
+    assert(_found_init, "OpaqueLoopInitNode must always be found");
+    return current->as_Bool();
+  }
+};
+
+// This class caches a single OpaqueLoopInitNode and OpaqueLoopStrideNode. If the node is not cached, yet, we clone it
+// and store the clone in the cache to be returned for subsequent calls.
+class CachedOpaqueLoopNodes {
+  OpaqueLoopInitNode* _cached_opaque_new_init;
+  OpaqueLoopStrideNode* _cached_new_opaque_stride;
+  PhaseIdealLoop* _phase;
+  Node* _new_ctrl;
+
+ public:
+  CachedOpaqueLoopNodes(PhaseIdealLoop* phase, Node* new_ctrl)
+      : _cached_opaque_new_init(nullptr),
+        _cached_new_opaque_stride(nullptr),
+        _phase(phase),
+        _new_ctrl(new_ctrl) {}
+
+  OpaqueLoopInitNode* clone_if_not_cached(OpaqueLoopInitNode* opaque_init) {
+    if (_cached_opaque_new_init == nullptr) {
+      _cached_opaque_new_init = _phase->clone_and_register(opaque_init, _new_ctrl)->as_OpaqueLoopInit();
+    }
+    return _cached_opaque_new_init;
+  }
+
+  OpaqueLoopStrideNode* clone_if_not_cached(OpaqueLoopStrideNode* opaque_stride) {
+    if (_cached_new_opaque_stride == nullptr) {
+      _cached_new_opaque_stride = _phase->clone_and_register(opaque_stride, _new_ctrl)->as_OpaqueLoopStride();
+    }
+    return _cached_new_opaque_stride;
+  }
+};
+
+// The transformations of this class clone the existing OpaqueLoop* nodes without any other update.
+class CloneOpaqueLoopNodes : public TransformOpaqueLoopNodes {
```

Comment:
Would it not be much simpler, if you just had one such Transform class, which can have both a new init and stride node as input, but if they are nullptr, then we just clone instead? It would also simplify the code in `clone_assertion_predicate_and_initialize`.